static/app/actionCreators/dashboards.tsx

```diff
@@ -40,18 +40,23 @@ export function createDashboard(
   newDashboard: DashboardDetails,
   duplicate?: boolean
 ): Promise<DashboardDetails> {
-  const {title, widgets} = newDashboard;
+  const {title, widgets, projects, environment, period, start, end} = newDashboard;
 
   const promise: Promise<DashboardDetails> = api.requestPromise(
     `/organizations/${orgId}/dashboards/`,
     {
       method: 'POST',
-      data: {title, widgets: widgets.map(widget => omit(widget, ['tempId'])), duplicate},
-      query: {
-        // TODO: This should be replaced in the future with projects
-        // when we save Dashboard page filters. This is being sent to
-        // bypass validation when creating or updating dashboards
-        project: [ALL_ACCESS_PROJECTS],
+      data: {
+        title,
+        widgets: widgets.map(widget => omit(widget, ['tempId'])),
+        duplicate,
+        // Ensure projects and environment are sent as arrays, or undefined in the request
+        // location.query will return a string if there's only one value
+        projects: typeof projects === 'string' ? [projects] : projects,
+        environment: typeof environment === 'string' ? [environment] : environment,
```

Comment:
The typing seems inconsistent here because TS expects `projects` to be a `number[]` and `environment` to be `string[] | undefined` but the code here is conditional on if either variable are of type `string`.
We're forcing `projects` and `environment` to be arrays here, but maybe we can pull this logic out of this function and put it in `detail.tsx`? That way we can reliable assume that `newModifiedDashboard` is a properly of `DashboardDetails` type.

Reply:
Good point. Pulling values off of `location.query` is always kind of annoying because they're typed as `any`. I've at least moved this logic upwards to where we call `createDashboard` and done the type checking there.

At the time it was a trade-off of putting the logic closest to where we know we should check something (i.e. destructuring off of `location.query`), or putting it at the point where we can ensure the correct state before the request. `createDashboard` is only called in one place so I agree with moving the logic up to `detail.tsx`